src/main/java/org/opendevstack/provision/model/OpenProjectData.java

```diff
@@ -121,8 +121,21 @@ public int hashCode() {
     if (quickstartersToRemove == null) {
       return this.quickstarters;
     }
-    for (Map<String, String> quickstarter : quickstartersToRemove) {
-      quickstarters.remove(quickstarter);
+    for (Map<String, String> quickStarterToRemove : quickstartersToRemove) {
+      if (quickStarterToRemove.get(COMPONENT_ID_KEY) == null) {
+        throw new NullPointerException("Cannot delete quickstarter with id null!");
+      }
+      // loop over the currently provisioned quickstarters, and find the
+      // one with similar id - to then remove it.
+      List<Map<String, String>> currentQuickstarters = getQuickstarters();
+      for (int i = 0; i < currentQuickstarters.size(); i++) {
+        Map<String, String> currentQuickstarter = currentQuickstarters.get(i);
+        if (currentQuickstarter
+            .get(COMPONENT_ID_KEY)
+            .equalsIgnoreCase(quickStarterToRemove.get(COMPONENT_ID_KEY))) {
+          quickstarters.remove(i);
+        }
+      }
     }
     return quickstarters;
```

Comment:
After a quick review I found some design issues with this class, i.e. it break encapsulation of internals by exposing the quickstarters map to the method caller. This could modify the quickstarters map outside the boundaries of the OpenProjectData class. Also removal of quickstarters is not thread safe. However, I'm concentrating here on fix for the reported bug...
I would suggest that we modify the method "removeQuickstartersFromProject" to not return the quickstarters map. Simply because the method name does make clear what is returning. It has currently only one caller which it doesn't need the quickstarters map. Also from the caller point of view is not clear if a list of the removed quickstarters or the remaining quickstarters are returned. A developer will need to review the implementation to understand what this method is returning. Alternatively we could rename the method to reflect its intend (e.g. "removeQuickstartersFromProjectAndReturnRemainingOnes") but I'm would prefer to remove the return clauses. Sounds good?

Reply:
@stitakis - I changed it now, to return the removed quickstarters. Happier?